File: backend/models/network.py
"""
Network topology data models (Pydantic/dataclass)
- Node properties, interface details, routing tables
- IP ownership and VLAN configurations
- Maps to Batfish API response structures
"""
from dataclasses import dataclass, field, asdict
from typing import Any
from .enums import NodeType


@dataclass
class InterfaceDetail:
    """Interface detailed information"""
    name: str
    admin_up: bool
    active: bool
    description: str = ""
    speed: int | None = None
    mtu: int = 1500
    vlan: int | None = None
    allowed_vlans: list[int] = field(default_factory=list)
    primary_address: str | None = None
    bandwidth: int | None = None
    inactive_reason: str | None = None


@dataclass
class VLANInfo:
    """VLAN information"""
    id: int
    name: str = ""
    interfaces_count: int = 0
    interfaces: list[str] = field(default_factory=list)
    vxlan_vni: int | None = None


@dataclass
class RouteEntry:
    """Routing table entry"""
    network: str
    next_hop: str
    next_hop_ip: str | None = None
    next_hop_interface: str | None = None
    protocol: str = "static"
    metric: int | None = None
    admin_distance: int | None = None
    tag: int | None = None


@dataclass
class NetworkNode:
    """Network node/device data class"""
    id: str
    name: str
    type: NodeType
    platform: str
    hostname: str = ""
    interfaces: list[InterfaceDetail] = field(default_factory=list)
    vlans: list[VLANInfo] = field(default_factory=list)
    routes: dict[str, int] = field(default_factory=dict)
    ip_addresses: list[str] = field(default_factory=list)
    protocols: dict[str, Any] = field(default_factory=dict)
    health_status: str = "healthy"
    config_issues: list[str] = field(default_factory=list)

    # Additional properties from nodeProperties (22種類のデータ対応)
    configuration_format: str | None = None
    dns_servers: list[str] = field(default_factory=list)
    domain_name: str | None = None
    ntp_servers: list[str] = field(default_factory=list)
    logging_servers: list[str] = field(default_factory=list)
    snmp_trap_servers: list[str] = field(default_factory=list)
    tacacs_servers: list[str] = field(default_factory=list)
    vrfs: list[str] = field(default_factory=list)
    zones: list[str] = field(default_factory=list)

    def to_dict(self) -> dict:
        """Convert to dictionary for JSON serialization"""
        data = asdict(self)
        data['type'] = self.type.value
        return data


@dataclass
class NetworkEdge:
    """Network edge/link data class"""
    source: str
    target: str
    source_port: str
    target_port: str
    source_ip: str | None = None
    target_ip: str | None = None
    protocol: str | None = None
    bandwidth: int | None = None
    utilization: float | None = None

    def to_dict(self) -> dict:
        """Convert to dictionary"""
        return asdict(self)

File: frontend/src/types/network.ts
/**
 * Network topology core type definitions
 * Node properties, interface details, routing tables, and IP ownership
 * Used throughout the application for network device modeling
 */
import { NodeType } from './enums'

export interface InterfaceDetail {
  name: string
  admin_up: boolean
  active: boolean
  description?: string
  speed?: number
  mtu: number
  vlan?: number
  allowed_vlans: number[]
  primary_address?: string
  bandwidth?: number
  inactive_reason?: string
}

export interface VLANInfo {
  id: number
  name?: string
  interfaces_count: number
  interfaces: string[]
  vxlan_vni?: number
}

export interface RouteEntry {
  network: string
  next_hop: string
  next_hop_ip?: string
  next_hop_interface?: string
  protocol: string
  metric?: number
  admin_distance?: number
  tag?: number
}

export interface NetworkNode {
  id: string
  name: string
  type: NodeType
  platform: string
  hostname?: string
  interfaces: InterfaceDetail[]
  vlans: VLANInfo[]
  routes: Record<string, number>
  ip_addresses: string[]
  protocols: Record<string, any>
  health_status: string
  config_issues: string[]

  configuration_format?: string
  dns_servers: string[]
  domain_name?: string
  ntp_servers: string[]
  logging_servers: string[]
  snmp_trap_servers: string[]
  tacacs_servers: string[]
  vrfs: string[]
  zones: string[]
}

export interface NetworkEdge {
  source: string
  target: string
  source_port: string
  target_port: string
  source_ip?: string
  target_ip?: string
  protocol?: string
  bandwidth?: number
  utilization?: number
}

export interface NodeProperties {
  node: string
  configuration_format?: string
  dns_servers: string[]
  dns_source_interface?: string
  domain_name?: string
  hostname: string
  ipsec_vpns: string[]
  interfaces: string[]
  logging_servers: string[]
  logging_source_interface?: string
  ntp_servers: string[]
  ntp_source_interface?: string
  snmp_source_interface?: string
  snmp_trap_servers: string[]
  tacacs_servers: string[]
  tacacs_source_interface?: string
  vendor: string
  vrfs: string[]
  zones: string[]
}

export interface InterfaceProperties {
  interface: string
  access_vlan?: number
  active: boolean
  admin_up: boolean
  all_prefixes: string[]
  allowed_vlans: string
  auto_state_vlan: boolean
  bandwidth?: number
  blacklisted: boolean
  channel_group?: string
  channel_group_members: string[]
  declared_names: string[]
  description: string
  encapsulation_vlan?: number
  hsrp_groups: any[]
  hsrp_version?: string
  incoming_filter_name?: string
  interface_type: string
  mtu: number
  native_vlan?: number
  ospf_area_name?: string
  ospf_cost?: number
  ospf_enabled: boolean
  ospf_network_type?: string
  ospf_passive: boolean
  outgoing_filter_name?: string
  primary_address?: string
  primary_network?: string
  proxy_arp: boolean
  rip_enabled: boolean
  rip_passive: boolean
  speed?: number
  switchport: boolean
  switchport_mode: string
  switchport_trunk_encapsulation: string
  vlan?: number
  vrf: string
  zone?: string
}

export interface Route {
  node: string
  vrf: string
  network: string
  next_hop: string
  next_hop_ip?: string
  next_hop_interface?: string
  protocol: string
  metric?: number
  admin_distance?: number
  tag?: number
}